### PR TITLE
[gatsby-source-nacelle] Add Support for Gatsby Image & Incremental Builds

### PR DESCRIPTION
**Pull request by NWRichmond**
### Jira Tasks

- [DRMS-800](https://nacelle.atlassian.net/browse/DRMS-800)
  - > Gatsby Integration
- [DRMS-802](https://nacelle.atlassian.net/browse/DRMS-802)
  - > Add Gatsby Image support to Product, Collection, and Content types
- [DRMS-822](https://nacelle.atlassian.net/browse/DRMS-822)
  - > Refactor to not use gatsby-graphql-toolkit
- [DRMS-823](https://nacelle.atlassian.net/browse/DRMS-823)
  - > Add support for incremental builds

### Type of Change

- [x] Non-breaking feature
- [x] Chore (docs, refactor, etc)

#### Incremental Builds

- Adds support for incremental builds
  - Leverages the `indexedAt` property of products, collections, and content to only update nodes which have been re-indexed since the last build
  - Provides a `cacheDuration` configuration parameter to allow cache busting after a prescribed interval
- Adds documentation for incremental build support

#### Gatsby Image

traced SVG images loading in an eCommerce storefront

- Adds support for [Gatsby Image](https://using-gatsby-image.gatsbyjs.org/) for products, collections, and content***
- Documents the use of Gatsby Image with `@nacelle/gatsby-source-nacelle`
- In `examples/gatsby`, demonstrates the use of [`gatsby-plugin-image`](https://www.npmjs.com/package/gatsby-plugin-image) with `@nacelle/gatsby-source-nacelle`
- Components used in `examples/next` are added to `examples/gatsby`

> ***NOTE: for content, only the `featuredMedia.src` is compatible with Gatsby Image. Nested images inside of a content entry's `fields` or `sections` are not supported, because `fields` and `sections` are of the `JSON` type and images held within them cannot be accounted for in the [schema](https://www.gatsbyjs.com/docs/schema-customization/).

#### Refactor to not use gatsby-graphql-toolkit

- Instead of using [`gatsby-graphql-source-toolkit`](https://www.npmjs.com/package/gatsby-graphql-source-toolkit) to source nodes from Hail Frequency, use the Nacelle SDK to source data
  - Manual node sourcing & schema customization was required for Content anyway; this refactor goes whole-hog on this approach and applies it to Product, Collection, and Space data
  - Improves code clarity (every aspect of data sourcing is transparent in the source code)
  - Prevents `.gql` fragment files from being copied into the user's project, which was a bit disorienting / potential source of confusion
- Allows users to opt out of using Gatsby Image
  - Removes `gatsby-source-filesystem` as a dependency, and adds it as `optionalDependency`
  - Updates docs

### How to Test

From the project root, `npm run bootstrap && cd examples/gatsby && npm run dev`

### CODE REVIEW

File: examples/gatsby/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "nacelle-gatsby-example",
-  "version": "2.8.1-alpha.0",
+  "version": "2.9.0-alpha.0",
```
Comment (johnstonbl01) [inline examples/gatsby/package.json:3]: Please make sure to just bump the version to `2.9.0` without the alpha before merging & publishing.

File: examples/gatsby/src/components/CollectionGrid/CollectionGrid.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          handle: collectionHandle
+        });
+        const products = await $nacelle.data.products({
+          handles: collection.productLists.find(
+            (listEntry) => listEntry.slug === 'default'
+          ).handles
+        });
+        setProducts(products);
+      } catch {
+        console.warn(`Collection not found with handle: '${collectionHandle}'`);
```
Comment (johnstonbl01) [inline examples/gatsby/src/components/CollectionGrid/CollectionGrid.js:24]: Theoretically this could be an issue with `$nacelle.data.products` as well, right? Am wondering if we want to make the error more generic and use the error message from the SDK? Unsure if that's less useful or not.

File: examples/gatsby/src/components/ProductCard/ProductCard.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import * as React from 'react';
```
Comment (johnstonbl01) [inline examples/gatsby/src/components/ProductCard/ProductCard.js:1]: We can probably just `import React from 'react'` here, no?

File: examples/gatsby/src/components/ProductGallery/ProductGallery.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
```
Comment (johnstonbl01) [inline examples/gatsby/src/components/ProductGallery/ProductGallery.js:1]: Similar to above -- how come we're importing `*`?

File: examples/gatsby/src/providers/ProductSearch.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
+import * as React from 'react';
```
Comment (johnstonbl01) [inline examples/gatsby/src/providers/ProductSearch.js:1]: These import lines are redundant.

```
import React, { useMemo, useContext } from 'react';`
```

File: packages/gatsby-source-nacelle/gatsby-node.js [hunk 6]
```diff
@@ -1,42 +1,95 @@
...
       `Toggle Contentful Preview on and off (IMPORTANT: requires that both 'contentfulPreviewSpaceId' and 'contentfulPreviewApiToken' are also set)`
+    ),
+    cacheDuration: Joi.number().description(
+      'Max duration in ms that gatsby-source-nacelle caches product, collection, and content data from previous builds'
     )
   });
 };
 
+exports.sourceNodes = async (gatsbyApi, pluginOptions) => {
+  await Promise.all([
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/gatsby-node.js:34]: Should we `catch` here?

File: packages/gatsby-source-nacelle/gatsby-node.js [hunk 7]
```diff
@@ -1,42 +1,95 @@
...
-  if (cmsPreviewEnabled(pluginOptions)) {
-    // source content data from Contentful Preview API & convert to Gatsby nodes
-    await sourceContentfulPreviewNodes(gatsbyApi, pluginOptions);
+  try {
+    gatsbySourceFilesystem = require('gatsby-source-filesystem');
+  } catch (err) {
+    gatsbySourceFilesystem = null;
   }
+
+  if (gatsbySourceFilesystem) {
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/gatsby-node.js:63]: Instead of putting this inside another `if` -- does it make sense to just do this within the `try` block? We can `return null` if an error is caught (i.e. the dependency doesn't exist).

File: packages/gatsby-source-nacelle/src/source-nodes/content.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  const client = nacelleClient({
+    previewMode: cmsPreviewEnabled(pluginOptions),
+    nacelleSpaceId,
+    nacelleGraphqlToken,
+    contentfulPreviewSpaceId,
+    contentfulPreviewApiToken
+  });
+
+  try {
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/src/source-nodes/content.js:28]: This looks very similar to `collections`. Is there an opportunity to abstract some of this into small functions that can be used across both data types?

File: packages/gatsby-source-nacelle/src/source-nodes/products.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  const { createNode, touchNode } = actions;
+
+  const { nacelleSpaceId, nacelleGraphqlToken } = pluginOptions;
+
+  const client = nacelleClient({
+    nacelleSpaceId,
+    nacelleGraphqlToken
+  });
+
+  try {
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/src/source-nodes/products.js:19]: Same as comment on `content.js`

File: packages/gatsby-source-nacelle/src/type-defs/index.js [hunk 10]
```diff
@@ -1,11 +1,133 @@
 module.exports = `
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/src/type-defs/index.js:1]: Were you not able to figure out a programmatic way to handle these updates to the schema? Just curious

File: packages/gatsby-source-nacelle/src/utils/createRemoteImageFileNode.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+) {
+  let createRemoteFileNode;
+  try {
+    createRemoteFileNode = require('gatsby-source-filesystem')
+      .createRemoteFileNode;
+  } catch (err) {
+    createRemoteFileNode = null;
+  }
+
+  if (createRemoteFileNode) {
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/src/utils/createRemoteImageFileNode.js:41]: Similar to previous comment -- should be do this inside the above `try` block rather than putting it inside another conditional. Additionally, I wonder if there's an opportunity to create some smaller functions here - this code can get hard to follow when you have a `try => if => try => if => if` in a single function, and would likely be easy to misunderstand / mess up in the future.

File: packages/gatsby-source-nacelle/src/utils/createRemoteImageFileNode.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      }
+    } catch (err) {
+      throw new Error(
+        `Problem creating file node for remote image: ${err.message}`
+      );
+    }
+  }
+}
+
+/**
```
Comment (johnstonbl01) [inline packages/gatsby-source-nacelle/src/utils/createRemoteImageFileNode.js:82]: Yay docs! 🙏 🎉

**Review by johnstonbl01 — CHANGES_REQUESTED** [review]
If we need to move fast on this, then some of these are nitpicks. I'm happy if we create stories to come back and address these later if needed.

File: examples/gatsby/package.json (hunk 1 above)
Comment (NWRichmond) [reply examples/gatsby/package.json:3]: Sure thing - I'll just select "minor release" from the Lerna publishing options.

File: packages/gatsby-source-nacelle/src/type-defs/index.js (hunk 10 above)
Comment (NWRichmond) [reply packages/gatsby-source-nacelle/src/type-defs/index.js:1]: I spent some time trying to solve this via remote schema introspection, but didn't get very far. It's worth revisiting later, for sure.

File: examples/gatsby/src/components/ProductGallery/ProductGallery.js (hunk 4 above)
Comment (NWRichmond) [reply examples/gatsby/src/components/ProductGallery/ProductGallery.js:1]: The `import * as React from 'react` pattern ([is used in TypeScript React projects](https://stackoverflow.com/a/55285991/6387812)). Since this isn't a TypeScript project, I'm happy to change it to `import React from 'react'` if it helps with consistency between example projects.

Comment (johnstonbl01) [reply examples/gatsby/src/components/ProductGallery/ProductGallery.js:1]: Yeah, none of our current React TS projects import in that way. Explicit imports > implicit exports IMO. It's also just a good pattern / practice to promote b/c it's required for tree-shaking to work (although pretty much all of React gets bundled anyway, that's not the case for other libraries).

File: packages/gatsby-source-nacelle/src/type-defs/index.js (hunk 10 above)
Comment (johnstonbl01) [reply packages/gatsby-source-nacelle/src/type-defs/index.js:1]: 👍  We can resolve this, but should create a story to come back and take care of this.

File: packages/gatsby-source-nacelle/src/source-nodes/content.js (hunk 8 above)
Comment (NWRichmond) [reply packages/gatsby-source-nacelle/src/source-nodes/content.js]: UPDATE: refactored to use IoC / DI to stay DRY

File: packages/gatsby-source-nacelle/src/source-nodes/products.js (hunk 9 above)
Comment (NWRichmond) [reply packages/gatsby-source-nacelle/src/source-nodes/products.js]: UPDATE: refactored to use IoC / DI to stay DRY

File: examples/gatsby/src/components/CollectionGrid/CollectionGrid.js (hunk 2 above)
Comment (NWRichmond) [reply examples/gatsby/src/components/CollectionGrid/CollectionGrid.js]: Good point! I've added some more explicit error handling.

**Review by johnstonbl01 — APPROVED** [review]
Please just make sure to update the version! 🥇
